src/manager/checkpoint.rs

```diff
@@ -0,0 +1,160 @@
+/// `(child_subnet, parent_subnet)`, where `child_subnet` is a subnet that we need to actively
+/// manage checkpoint for. This means that for each `child_subnet` there exists at least one account
+/// for which we need to submit checkpoints on behalf of to `parent_subnet`, which must also be
+/// present in the map.
+fn subnets_to_manage(subnets: HashMap<String, Subnet>) -> Vec<(Subnet, Subnet)> {
+    // First, we remap subnets by SubnetID.
+    let subnets_by_id: HashMap<SubnetID, Subnet> = subnets
+        .values()
+        .map(|s| (s.id.clone(), s.clone()))
+        .collect();
+
+    // Then, we filter for subnets that have at least one account and for which the parent subnet
+    // is also in the map, and map into a Vec of (child_subnet, parent_subnet) tuples.
+    subnets_by_id
+        .values()
+        .filter(|s| !s.accounts.is_empty())
+        .filter(|s| s.id.parent().is_some() && subnets_by_id.contains_key(&s.id.parent().unwrap()))
+        .map(|s| (s.clone(), subnets_by_id[&s.id.parent().unwrap()].clone()))
+        .collect()
+}
+
+/// Monitors a subnet `child` for checkpoint blocks. It emits an event for every new checkpoint block.
+async fn manage_subnet(child: Subnet, parent: Subnet) -> Result<()> {
+    let child_client = LotusJsonRPCClient::from_subnet(&child);
+    let parent_client = LotusJsonRPCClient::from_subnet(&parent);
+
+    // Read the parent's chain head and obtain the tip set CID.
+    let parent_head = parent_client.chain_head().await?;
+    let cid_map = parent_head.cids.last().unwrap().clone();
+    let parent_tip_set = Cid::try_from(cid_map)?;
+
+    // Extract the checkpoint period from the state of the subnet actor in the parent.
+    let state = parent_client
+        .ipc_read_subnet_actor_state(parent_tip_set)
+        .await?;
+    let period = state.check_period;
+
+    // We should have a way of knowing whether the validator has voted in the current open
+    // checkpoint epoch. For now, we maintain this information in memory and upon a restart only
+    // vote for the next epoch.
+
+    // We can now start looping. In each loop we read the child subnet's chain head and check if
+    // it's a checkpoint epoch. If it is, we construct and submit a checkpoint.
+    loop {
+        let child_head = child_client.chain_head().await?;
+        let epoch: ChainEpoch = ChainEpoch::try_from(child_head.height)?;
+        if epoch % period == 0 {
+            // It's a checkpointing epoch and we may have checkpoints to submit.
+
+            // First, we check which accounts are in the validator set. This is done by reading
+            // the parent's chain head and requesting the state at that tip set.
+            let parent_head = parent_client.chain_head().await?;
+            // A key assumption we make now is that each block has exactly one tip set. We panic
+            // if this is not the case as it violates our assumption.
+            assert_eq!(parent_head.cids.len(), 1);
```

Comment:
This is the case for Mir, once we move to Filecoin mainnet as the root for M3 this won't be the case. Let's maybe add a TODO here to remind ourselves. Thanks.

Reply:
I'll add a TODO. The reason I set this panic is precisely because it should blow up to reminds that the logic here needs to be updated.

Reply:
SGTM. I like the approach.